### PR TITLE
UCT/SM/TEST: Add wakeup support for shared memory

### PR DESCRIPTION
**Pull request by yosefe**
# Why
Allow applications which require wakeup take advantage of shared memory.
The idea is that the receiver sets a "signal" flag in FIFO head pointer and makes sure the next sender will observe this bit.
The sender will signal on Unix-domain socket if the signal bit is set.

### CODE REVIEW

File: src/uct/sm/scopy/base/scopy_iface.c [hunk 1]
```diff
@@ -136,6 +139,19 @@ unsigned uct_scopy_iface_progress(uct_iface_h tl_iface)
     return count;
 }
 
+ucs_status_t uct_scopy_iface_event_arm(uct_iface_h tl_iface, unsigned events)
```
Comment (dmitrygx) [inline src/uct/sm/scopy/base/scopy_iface.c:142]: if the user requested waiting for events and then does PUT/GET operation, the callback won't be called upon completion.
is it correct behavior?

Comment (yosefe) [reply src/uct/sm/scopy/base/scopy_iface.c:142]: yes, the user should arm with success before going to sleep

**Comment by yosefe** [comment]
@dmitrygx @brminich can you pls review?

**Review by brminich — APPROVED** [review]

**Review by dmitrygx — APPROVED** [review]

**Comment by yosefe** [comment]
BTW, after this PR we can remove proxy_ep from UCP

**Review by brminich — APPROVED** [review]

**Comment by yosefe** [comment]
@brminich @hoopoepg @dmitrygx can you pls look again, this PR was failing on test_uct_event, so needed to set higher "spurious event threshold" when flow-control is enabled. This 1st commit adds gtest infra to skip a test if some configuration param is not present by "NAME:=VALUE", before we could only fail or ignore.

File: test/gtest/uct/test_event.cc [hunk 2]
```diff
@@ -166,20 +170,38 @@ void test_uct_event::test_recv_am(unsigned arm_flags, unsigned send_flags)
...
+                     !check_event_caps(UCT_IFACE_FLAG_EVENT_RECV),
+                     "RC_FC_ENABLE?=n")
 {
-    test_recv_am(UCT_EVENT_RECV, 0);
+    test_recv_am(UCT_EVENT_RECV, 0, 0.02);
+}
+
+/*
+ * Enable flow control (skip the test if transport does not support it), and
+ * expect at most 10% spurious  events.
```
Comment (brminich) [inline test/gtest/uct/test_event.cc:188]: extra whitespace before `events`

File: test/gtest/uct/test_event.cc [hunk 3]
```diff
@@ -166,20 +170,38 @@ void test_uct_event::test_recv_am(unsigned arm_flags, unsigned send_flags)
     free(recv_buffer);
 }
 
-UCS_TEST_SKIP_COND_P(test_uct_event, am,
+/*
+ * Disable flow control (if transport supports it), and expect at most 2%
+ * spurious events.
```
Comment (brminich) [inline test/gtest/uct/test_event.cc:175]: what are those 2%?

File: src/uct/sm/scopy/base/scopy_iface.c [hunk 4]
```diff
@@ -77,6 +77,9 @@ void uct_scopy_iface_query(uct_scopy_iface_t *iface, uct_iface_attr_t *iface_att
                                           UCT_IFACE_FLAG_PUT_ZCOPY |
                                           UCT_IFACE_FLAG_PENDING   |
                                           UCT_IFACE_FLAG_CONNECT_TO_IFACE;
+    iface_attr->cap.event_flags         = UCT_IFACE_FLAG_EVENT_SEND_COMP |
+                                          UCT_IFACE_FLAG_EVENT_RECV |
```
Comment (dmitrygx) [inline src/uct/sm/scopy/base/scopy_iface.c:81]: [minor] align by `|`

File: test/gtest/common/test.cc [hunk 5]
```diff
@@ -70,13 +70,36 @@ void test_base::set_config(const std::string& config_str)
...
-    if ((name.length() > 0) && name.at(name.length() - 1) == '?') {
-        name = name.substr(0, name.length() - 1);
-        optional = true;
+    mode = FAIL_IF_NOT_EXIST;
+
+    /*
+     * What happens if NAME is not a valid configuration key?
+     * - "NAME=VALUE"   : fail the test
+     * - "NAME?=VALUE"  : set UCX_NAME environment variable
+     * - "NAME:=VALUE"  : skip the test
```
Comment (dmitrygx) [inline test/gtest/common/test.cc:79]: maybe better `~=` or smth different?
IMO, `:=` means assignment and could be confusing then

File: test/gtest/uct/uct_test.cc [hunk 6]
```diff
@@ -1454,7 +1454,9 @@ bool uct_test::async_event_ctx::wait_for_event(entity &e, int timeout) {
         }
     }
 
-    int ret = poll(&wakeup_fd, 1, timeout);
+    int timeout_ms = static_cast<int>((timeout_sec * UCS_MSEC_PER_SEC) *
+                                      ucs::test_time_multiplier());
+    int ret = poll(&wakeup_fd, 1, timeout_ms);
```
Comment (dmitrygx) [inline test/gtest/uct/uct_test.cc:1459]: [minor] align by `=`

File: test/gtest/uct/test_event.cc (hunk 2 above)
Comment (yosefe) [reply test/gtest/uct/test_event.cc]: removed the comment, not relevant anymore

File: test/gtest/uct/test_event.cc (hunk 3 above)
Comment (yosefe) [reply test/gtest/uct/test_event.cc]: removed the comment, not relevant anymore

File: src/uct/sm/scopy/base/scopy_iface.c (hunk 4 above)
Comment (yosefe) [reply src/uct/sm/scopy/base/scopy_iface.c]: done

File: test/gtest/uct/uct_test.cc (hunk 6 above)
Comment (yosefe) [reply test/gtest/uct/uct_test.cc]: done

File: test/gtest/common/test.cc (hunk 5 above)
Comment (yosefe) [reply test/gtest/common/test.cc]: done

**Comment by dmitrygx** [comment]
looks like related
```
[ RUN      ] ud_verbs/test_uct_event.am_no_fc/1 <ud_verbs/mlx5_0:1>
[     INFO ] Testing component: ib
/scrap/azure/agent-03/AZP_WORKSPACE/2/s/contrib/../test/gtest/uct/test_event.cc:157: Failure
Value of: spurious_count
  Actual: 1
Expected: 0u
Which is: 0
[  FAILED  ] ud_verbs/test_uct_event.am_no_fc/1, where GetParam() = ud_verbs/mlx5_0:1 (2465 ms)
```

**Review by dmitrygx — APPROVED** [review]

**Review by dmitrygx — APPROVED** [review]

**Review by dmitrygx — APPROVED** [review]

**Pull request merged**